include/pybind11/cast.h

```diff
@@ -2198,6 +2198,17 @@ object object_api<Derived>::call(Args &&...args) const {
 
 PYBIND11_NAMESPACE_END(detail)
 
+
+/** \ingroup python_builtins
+    \rst
+    Return the registered type object for a C++ class, given as a template parameter.
+    py::type<T>() returns the Python type object previously registered for T.
+\endrst */
+template<typename T>
+handle type() {
```

Comment:
Should this be `handle` or `object`? I have a feeling that often, most user-facing cases return `object` rather than `handle`, or not?

Reply:
Yeah, from a quick perusal, that's what I generally see too.
FWIW, the pattern seems to be return `object`, but take `handle` as an argument.

Reply:
Yes, I believe a `PyObject*` has an implicit conversion to `handle`, so it's +- compatible with some raw C API type, but given that we manage refcounting for the user, returning `object` makes sense. If the user needs a raw `PyObject*` again, `release()` is there?

Reply:
Is it possible to delete a user-defined type? Refcounting seemed to be unnecessary if it's something that can't be deleted.

Reply:
That is also true, yes. Yes, I think these types can go out of scope, but pybind11 breaks at that point, anyway ;-) 
But then we should document how to convert it to `py::object` if needed (I guess this should be a `reinterpret_borrowed`?) Are there other functions that return a `handle` in such case that aren't in the `detail` namespace, though?

My use case now also works with `handle` though, so maybe this is good enough, yes.

```c++
template <typename Mixin>
void addMixinBase(py::handle &binding) {
	auto typeHandle = py::detail::get_type_handle(typeid(Mixin), true);
	binding.attr("__bases__") = py::make_tuple(typeHandle) + binding.attr("__bases__");
}
```

Reply:
> but pybind11 breaks at that point, anyway ;-)

In case anyone's interested, I recently worked out this example, in the context of #2335:

<details>
  <summary>Hiding this example because it's too long/off-topic</summary>

```c++
#include <pybind11/pybind11.h>

namespace py = pybind11;

class X {};
class Y {};

PYBIND11_MODULE(example, m)
{
        py::class_<X>(m, "X");
        py::class_<Y>(m, "Y");
        m.def("foo", []() { return X(); });
        m.def("bar", []() { return Y(); });
}
```

```pycon
>>> import example
>>> example.X
<class 'example.X'>
>>> example.Y
<class 'example.Y'>
>>> example.foo()
<example.X object at 0x7f7987a83c70>
>>> example.bar()
<example.Y object at 0x7f7987a83d70>
>>> import gc
>>> gc.collect()
0
>>> gc.collect()
0
>>> gc.collect()
0
>>> example.X = example.Y
>>> example.X
<class 'example.Y'>
>>> gc.collect()
0
>>> gc.collect()
0
>>> gc.collect()
0
>>> gc.get_referrers(type(example.foo()))
[{'__name__': 'example', '__doc__': None, '__package__': None, '__loader__': None, '__spec__': None, 'X': <class 'example.X'>, 'Y': <class 'example.Y'>, 'foo': <built-in method foo of PyCapsule object at 0x7f7987a80600>, 'bar': <built-in method bar of PyCapsule object at 0x7f7987a80660>, '__file__': '/home/yannick/tmp/sandbox/build/example.cpython-38-x86_64-linux-gnu.so'}, (<class 'example.X'>, <class 'pybind11_builtins.pybind11_object'>, <class 'object'>), <slot wrapper '__init__' of 'example.X' objects>]
>>> del gc.get_referrers(type(example.foo()))[0]['X']
>>> gc.collect()
0
>>> gc.collect()
5
>>> gc.collect()
0
>>> gc.collect()
0
>>> example.bar()
<example.Y object at 0x7f7987a02ef0>
>>> example.foo()
<example.X object at 0x7f7987a02e70>
>>> example.foo()
<example.X object at 0x7f7987a02ef0>
>>> 42
Segmentation fault (core dumped)
```
</details>

Reply:
I'm not against it being an object, this was just my thinking for handle. It works for my use case, too:

```cpp
.def_property_readonly_static(
    "_storage_type",
    [](py::object) {
        return py::detail::get_type_handle(
            typeid(typename histogram_t::storage_type), true);
    });
```

Reply:
`py::object_api<T>::get_type()` also returns a `handle` and is public. I'm OK with keeping it like this, then, I think.